node/src/actors/chain_manager/handlers.rs

```diff
@@ -178,6 +184,36 @@ impl Handler<EpochNotification<EveryEpochPayload>> for ChainManager {
                                 "There was no valid block candidate to consolidate for epoch {}",
                                 previous_epoch
                             );
+                            if get_protocol_version(Some(previous_epoch)) == ProtocolVersion::V2_0 {
+                                let rank_subset: Vec<_> = stakes
+                                    .rank(Capability::Mining, previous_epoch)
+                                    .take(MINING_REPLICATION_FACTOR)
+                                    .map(|sk| sk)
+                                    .collect();
+                                let mut max_power = 0;
+                                for (stake_key, power) in rank_subset.into_iter() {
+                                    // Initialize max_power if necessary
+                                    if max_power == 0 {
+                                        max_power = power;
+                                    };
+
+                                    // Only slash validators which should have proposed a block with 100% certainty
+                                    if power.saturating_mul(
+                                        MINING_REPLICATION_FACTOR.try_into().unwrap(),
+                                    ) >= max_power
+                                    {
+                                        log::warn!(
+                                            "Slashed the power of {} as it did not propose a block",
+                                            stake_key.validator
+                                        );
+                                        let _ = stakes.reset_age(
+                                            stake_key.validator,
+                                            Capability::Mining,
+                                            msg.checkpoint,
+                                        );
+                                    }
+                                }
+                            }
```

Comment:
In case of no valid block candidate to consolidate for previous epoch, the top MINING_REPLICATION_FACTOR eligible validators should be slashed, whereas with this code only those with at least 1 / MINING_REPLICATION_FACTOR of the highest power would actually get slashed.

Reply:
Besides, I'd rather recommend to reset coin age in proportion to each one's rank position:
- First: reset coin age to zero
- Second: subtract half the current age
- Third: subtract one third of current age 
- And so on

This way we would avoid/minimize the chances of a colluding set of big stakers to end up being within the same eligible set on periodically base.

Reply:
> In case of no valid block candidate to consolidate for previous epoch, the top MINING_REPLICATION_FACTOR eligible validators should be slashed, whereas with this code only those with at least 1 / MINING_REPLICATION_FACTOR of the highest power would actually get slashed.

Yes, my initial implementation actually slashed all validator up to `rf` validators. The problem with that is that you can actually be slashing validators which were not eligible to propose a block which seems rather unfair.

For example, imagine validator 1 and 2 have a power of 100k and validators 3 and 4 have a power of 10k. The former two are guaranteed to propose a block, but for the latter two it's a matter of whether their VRF is smaller than a target hash. Hence the choice in above algorithm to not slash validators 3 and 4. I'm up for changing that though.

Reply:
> Besides, I'd rather recommend to reset coin age in proportion to each one's rank position:
> 
> * First: reset coin age to zero
> * Second: subtract half the current age
> * Third: subtract one third of current age
> * And so on

Assume validator 1 and 2 have the same power, do we rest both to zero? I'm not sure how deterministic the rank function is with respect to multiple distinct nodes since it seems to be sorted by power only.

> This way we would avoid/minimize the chances of a colluding set of big stakers to end up being within the same eligible set on periodically base.

I understand the point and actually agree that this might be a good defense, but I do wonder if it's worse to have a single epoch where multiple colluding attackers are executing an attack (and hence we only have a single epoch with a block missing) or several epochs where those attackers separately try to run a griefing attack (and we might have several epochs with missing blocks if network connectivity is not perfect or there is also censoring taking place).

Reply:
> For example, imagine validator 1 and 2 have a power of 100k and validators 3 and 4 have a power of 10k. The former two are guaranteed to propose a block, but for the latter two it's a matter of whether their VRF is smaller than a target hash. Hence the choice in above algorithm to not slash validators 3 and 4. I'm up for changing that though.

According to the simulation source code that it's being replicated, with a replication factor of 4, validators 1 to 4 would indeed be eligible and expected to broadcast a valid block proposal. The VRF is just used for the other nodes to agree upon which to ultimately accept as best proposal, meaning that although validators 3 and 4 have less chances to win, they do have some, and further more, they are obliged to broadcast a block candidate. Therefore, in case of no blocks candidates in previous epoch, the 4 eligible validators should be staked, no matter if their absolute mining power.

Reply:
> Assume validator 1 and 2 have the same power, do we rest both to zero? I'm not sure how deterministic the rank function is with respect to multiple distinct nodes since it seems to be sorted by power only.

Yes, I do believe that eligible validators coincidentally having same power, they should still be slashed in different proportions. The easiest and most convenient way to implement that would be to rely on the implementation of  the `rank` function, as long as it is deterministic (i.e. non dependent on underlying OS, crate version, or thinks alike), as I believe it must be.

Reply:
> I understand the point and actually agree that this might be a good defense, but I do wonder if it's worse to have a single epoch where multiple colluding attackers are executing an attack (and hence we only have a single epoch with a block missing) or several epochs where those attackers separately try to run a griefing attack (and we might have several epochs with missing blocks if network connectivity is not perfect or there is also censoring taking place).

Considering that we're planning a replication factor as low as 4, I think it would be best to reduce the chances of multiple colluding attackers within same epoch. In fact, with this goal in mind, I'd rather suggest to consider a higher replication factor, like 8 or even 16, as long as we were certain that p2p gossiping doesn't grow to an unmanageable number of messages for the average node specs.

Reply:
> According to the simulation source code that I believe it's being replicated, with a replication factor of 4, validators 1 to 4 would indeed be eligible and expected to broadcast a valid block proposal. The VRF is just used for the other nodes to agree upon which to ultimately accept as best proposal, meaning that although validators 3 and 4 have less chances to win, they do still have some (even if higher ranked validators actually broadcast valid proposals), and further more, they are obliged to broadcast a block candidate whatsoever. Therefore, in case of no blocks candidates in previous epoch, the 4 eligible validators on that epoch should all be slashed, no matter if their absolute mining power is "low" compared to the others.

That is certainly not the case for the simulation and not how it is implemented in this PR. If a validator has more than `rf` less power, [this line](https://github.com/drcpu-github/proof-of-stake-tools/blob/2ed35795d4487b12bc6fe6bb0fedc52d18e9a6f3/simulate_block_mining.py#L149) combined with a node's VRF can result in them not proposing a block. It's not just that they have less chances to win, they will not even propose and broadcast a block.

> The easiest and most convenient way to implement that would be to rely on the implementation of the rank function, as long as it is deterministic (i.e. non dependent on underlying OS, crate version, or thinks alike), as I believe it must be.

I am simply not sure (the current implementation of rank](https://github.com/witnet/witnet-rust/blob/5ec6673c98f0a1f34553002d309d9945508a9033/data_structures/src/staking/stakes.rs#L232) is deterministic. I don't know if a BTreeMap in Rust maintains the order of insertion when iterated over or not and because we are only sorting on a single metric which can be the same for different validators, I imagine it could lead to non-determinism. Of course, that should be easily fixable by changing the line to something like this: `sorted_by_key(|(stake_key, power)| (stake_key.validator.clone(), *power))`

> In fact, with this goal in mind, I'd rather suggest to consider a higher replication factor, like 8 or even 16, as long as we were certain that p2p gossiping doesn't grow to an unmanageable number of messages for the average node specs.

I guess it's fine to increase the `rf` for the simple reason that the p2p traffic should reduce more or less exponentially. Any node will only forward a block if he considers it as a better block and hence most nodes will never see `rf` blocks. 8 feels fine to me, 16 feels a bit over the top.

Reply:
> That is certainly not the case for the simulation and not how it is implemented in this PR. If a validator has more than `rf` less power, [this line](https://github.com/drcpu-github/proof-of-stake-tools/blob/2ed35795d4487b12bc6fe6bb0fedc52d18e9a6f3/simulate_block_mining.py#L149) combined with a node's VRF can result in them not proposing a block. It's not just that they have less chances to win, they will not even propose and broadcast a block. Whether a block wins or not ultimately comes down to the selector used in the simulation, but in this PR power wins, then VRF, then block hash.

Sorry, I think I got a little lost here. You were basically saying that the top `rf` stakers cannot be slashed because there may be circumstances under which some of them would never "dare" to propose a block. My point is that, despite of whatever was intended or actually implemented in the simulator, isn't the whole point of the replication factor to always force a determined number of identities (no more, no less) to propose block candidates based on their power rank, no matter the relative power among the top `rf` candidates??

So in your above mentioned example, you said:

> For example, imagine validator 1 and 2 have a power of 100k and validators 3 and 4 have a power of 10k. The former two are guaranteed to propose a block, but for the latter two it's a matter of whether their VRF is smaller than a target hash. Hence the choice in above algorithm to not slash validators 3 and 4. I'm up for changing that though.

But I do believe that validators 1 to 4 must all propose and broadcast valid block candidates. True, some of them may refrain from doing so in case they detect a valid block being broadcasted from a eligible candidate with a higher power (or same power but a higher vrf). But still, all `rf` candidates SHOULD be prone and ready to broadcast asap their block proposal whatsoever. Because if none of them happened to do so, they MUST all get slashed.

Reply:
Yes, we can certainly require that every node is eligible to build and broadcast a block regardless of the power distribution for the top `rf` candidates. Some of the patches here are a 100% literal translation from the simulation though. I don't really see any downside to removing the VRF eligibility check (and modify the slashing accordingly) because it will almost never matter.

@aesedepece Before I correct this, what's your opinion?

Reply:
Moreover. Slashing ALL eligible candidates if no block gets consolidated on previous epoch, while NECESSARY, it would not yet be enough. As to minimize the chances of multiple identities colluding to perform a griefing attack (all such identities ranked within the top `rf` stakers), we should definitely consider a way to immediately slash reluctant candidates on the very same epoch on which the best block proposal gets consolidated. 

For instance, if the consolidated block proposal ultimately belongs to the `k`-est staker out of `rf` candidates, candidates ranked from `#1` to `#k-1` should be slashed immediately, whereas stakers `#k+1` to `rf` would not be slashed. This idea should be discussed and elaborated and eventually implemented on a different PR, of course.